File: frontend/lib/services/storage/supabase_storage_service.dart
import 'dart:io';
import 'package:supabase_flutter/supabase_flutter.dart';
import '../../config/supabase_config.dart';
import '../service_locator.dart';

class UploadResult {
  final bool success;
  final String? url;
  final String? error;

  UploadResult({
    required this.success,
    this.url,
    this.error,
  });
}

class SupabaseStorageService {
  // Backend ile uyumlu bucket isimleri kullan
  static const String _profileBucket = SupabaseConfig.profilePicturesBucket;
  static const String _coverBucket = SupabaseConfig.coverPicturesBucket;
  static const String _eventsBucket = SupabaseConfig.eventsPicturesBucket;
  static const String _groupsBucket = SupabaseConfig.groupsProfilePicturesBucket;
  static const String _postsBucket = SupabaseConfig.groupPostsImagesBucket;
  static const String _bikesBucket = SupabaseConfig.bikesImagesBucket;

  /// Profil fotoğrafı yükleme
  Future<UploadResult> uploadProfilePicture(File imageFile) async {
    try {
      print('🔥 SUPABASE PROFILE UPLOAD START');
      
      // Bucket'ın var olup olmadığını kontrol et
      String bucketToUse = _profileBucket;
      try {
        final buckets = await Supabase.instance.client.storage.listBuckets();
        print('🔥 Available buckets for profile: ${buckets.map((b) => b.name).toList()}');
        
        final profileBucketExists = buckets.any((bucket) => bucket.name == _profileBucket);
        if (!profileBucketExists) {
          print('🔥 WARNING: Profile bucket $_profileBucket does not exist!');
          
          // Alternatif bucket'ları dene
          final alternatives = ['moto-app-storage', 'cover_pictures', 'images', 'storage'];
          for (final altBucket in alternatives) {
            if (buckets.any((bucket) => bucket.name == altBucket)) {
              bucketToUse = altBucket;
              print('🔥 Using alternative bucket for profile: $bucketToUse');
              break;
            }
          }
          
          // Hiç bucket yoksa hata döndür
          if (!buckets.any((bucket) => bucket.name == bucketToUse)) {
            return UploadResult(
              success: false,
              error: 'Hiçbir storage bucket bulunamadı. Lütfen yönetici ile iletişime geçin.',
            );
          }
        }
      } catch (bucketError) {
        print('🔥 Profile bucket check error: $bucketError');
      }
      
      // Kullanıcı ID'sini al
      final userId = await _getCurrentUserId();
      if (userId == null) {
        throw Exception('Kullanıcı ID bulunamadı');
      }
      
      // Eski profil fotoğrafını sil
      await _deleteOldProfilePicture(userId, bucketToUse);
      
      final timestamp = DateTime.now().millisecondsSinceEpoch;
      final extension = imageFile.path.split('.').last;
      final fileName = 'profile_${timestamp}.$extension';
      final filePath = 'users/$userId/$fileName';
      
      print('🔥 Profile file path: $filePath');
      print('🔥 User ID: $userId');
      
      final response = await Supabase.instance.client.storage
          .from(bucketToUse)
          .upload(filePath, imageFile);

      print('🔥 Profile upload response: $response');

      final publicUrl = Supabase.instance.client.storage
          .from(bucketToUse)
          .getPublicUrl(filePath);

      print('🔥 Profile public URL: $publicUrl');

      return UploadResult(
        success: true,
        url: publicUrl,
      );
    } catch (e) {
      print('🔥 SUPABASE PROFILE UPLOAD ERROR: $e');
      return UploadResult(
        success: false,
        error: e.toString(),
      );
    }
  }


  /// Supabase bağlantısını test et
  Future<void> testSupabaseConnection() async {
    try {
      print('🔥 TESTING SUPABASE CONNECTION...');
      print('🔥 Supabase URL: ${SupabaseConfig.supabaseUrl}');
      print('🔥 Supabase Key: ${SupabaseConfig.supabaseAnonKey.substring(0, 20)}...');
      
      // Storage bucket'larını test et
      try {
        final buckets = await Supabase.instance.client.storage.listBuckets();
        print('🔥 Storage buckets test: SUCCESS');
        print('🔥 Available buckets: ${buckets.map((b) => b.name).toList()}');
      } catch (storageError) {
        print('🔥 Storage buckets test: FAILED - $storageError');
      }
      
      // Basit bir test yap (users tablosu yerine auth test)
      try {
        final user = Supabase.instance.client.auth.currentUser;
        print('🔥 Auth test: SUCCESS - User: ${user?.email ?? "Anonymous"}');
      } catch (authError) {
        print('🔥 Auth test: FAILED - $authError');
      }
    } catch (e) {
      print('🔥 General connection test: FAILED - $e');
    }
  }

  /// Kapak fotoğrafı yükleme
  Future<UploadResult> uploadCoverPicture(File imageFile) async {
    try {
      print('🔥 SUPABASE COVER UPLOAD START');
      print('🔥 Bucket: $_coverBucket');
      print('🔥 File path: ${imageFile.path}');
      print('🔥 File size: ${await imageFile.length()} bytes');
      
      // Supabase bağlantısını test et
      await testSupabaseConnection();
      
      // Hardcoded bucket kullan (RLS sorunu nedeniyle)
      String bucketToUse = 'cover_pictures'; // Supabase'de mevcut bucket
      print('🔥 Using hardcoded bucket: $bucketToUse');
      
      // Kullanıcı ID'sini al
      final userId = await _getCurrentUserId();
      if (userId == null) {
        throw Exception('Kullanıcı ID bulunamadı');
      }
      
      // Eski kapak fotoğrafını sil
      await _deleteOldCoverPicture(userId, bucketToUse);
      
      final timestamp = DateTime.now().millisecondsSinceEpoch;
      final extension = imageFile.path.split('.').last;
      final fileName = 'cover_${timestamp}.$extension';
      final filePath = 'users/$userId/$fileName';
      
      print('🔥 File path: $filePath');
      print('🔥 User ID: $userId');
      
      print('🔥 Attempting upload to bucket: $bucketToUse');
      
      final response = await Supabase.instance.client.storage
          .from(bucketToUse)
          .upload(filePath, imageFile);

      print('🔥 Upload response: $response');

      final publicUrl = Supabase.instance.client.storage
          .from(bucketToUse)
          .getPublicUrl(filePath);

      print('🔥 Public URL: $publicUrl');

      return UploadResult(
        success: true,
        url: publicUrl,
      );
    } catch (e) {
      print('🔥 SUPABASE COVER UPLOAD ERROR: $e');
      return UploadResult(
        success: false,
        error: e.toString(),
      );
    }
  }

  /// Post resmi yükleme
  Future<UploadResult> uploadPostImage(File imageFile) async {
    try {
      final timestamp = DateTime.now().millisecondsSinceEpoch;
      final extension = imageFile.path.split('.').last;
      final fileName = 'post_${timestamp}.$extension';
      
      final response = await Supabase.instance.client.storage
          .from(_postsBucket)
          .upload(fileName, imageFile);

      final publicUrl = Supabase.instance.client.storage
          .from(_postsBucket)
          .getPublicUrl(fileName);

      return UploadResult(
        success: true,
        url: publicUrl,
      );
    } catch (e) {
      return UploadResult(
        success: false,
        error: e.toString(),
      );
    }
  }

  /// Etkinlik resmi yükleme
  Future<UploadResult> uploadEventPicture(File imageFile) async {
    try {
      print('🔥 SUPABASE EVENT UPLOAD START');
      
      // Mevcut event klasörlerini kontrol et ve yeni numaralı klasör oluştur
      final nextFolderNumber = await _getNextEventFolderNumber();
      print('🔥 Next event folder number: $nextFolderNumber');
      
      final timestamp = DateTime.now().millisecondsSinceEpoch;
      final extension = imageFile.path.split('.').last;
      final fileName = 'event_${timestamp}.$extension';
      
      // Dosya yolu: events/{folder_number}/{file_name}
      final filePath = 'events/$nextFolderNumber/$fileName';
      print('🔥 Event file path: $filePath');
      
      final response = await Supabase.instance.client.storage
          .from(_eventsBucket)
          .upload(filePath, imageFile);

      print('🔥 Event upload response: $response');

      final publicUrl = Supabase.instance.client.storage
          .from(_eventsBucket)
          .getPublicUrl(filePath);

      print('🔥 Event public URL: $publicUrl');

      return UploadResult(
        success: true,
        url: publicUrl,
      );
    } catch (e) {
      print('🔥 SUPABASE EVENT UPLOAD ERROR: $e');
      return UploadResult(
        success: false,
        error: e.toString(),
      );
    }
  }

  /// Grup resmi yükleme
  Future<UploadResult> uploadGroupPicture(File imageFile) async {
    try {
      print('🔥 SUPABASE GROUP UPLOAD START');
      
      // Mevcut grup klasörlerini kontrol et ve yeni numaralı klasör oluştur
      final nextFolderNumber = await _getNextGroupFolderNumber();
      print('🔥 Next group folder number: $nextFolderNumber');
      
      final timestamp = DateTime.now().millisecondsSinceEpoch;
      final extension = imageFile.path.split('.').last;
      final fileName = 'group_${timestamp}.$extension';
      
      // Dosya yolu: groups/{folder_number}/{file_name}
      final filePath = 'groups/$nextFolderNumber/$fileName';
      print('🔥 Group file path: $filePath');
      
      final response = await Supabase.instance.client.storage
          .from(_groupsBucket)
          .upload(filePath, imageFile);

      print('🔥 Group upload response: $response');

      final publicUrl = Supabase.instance.client.storage
          .from(_groupsBucket)
          .getPublicUrl(filePath);

      print('🔥 Group public URL: $publicUrl');

      return UploadResult(
        success: true,
        url: publicUrl,
      );
    } catch (e) {
      print('🔥 SUPABASE GROUP UPLOAD ERROR: $e');
      return UploadResult(
        success: false,
        error: e.toString(),
      );
    }
  }

  /// Dosya silme
  Future<bool> deleteFile(String fileName, String bucketType) async {
    try {
      String bucketName;
      switch (bucketType) {
        case 'profile':
          bucketName = _profileBucket;
          break;
        case 'cover':
          bucketName = _coverBucket;
          break;
        case 'events':
          bucketName = _eventsBucket;
          break;
        case 'groups':
          bucketName = _groupsBucket;
          break;
        case 'posts':
          bucketName = _postsBucket;
          break;
        case 'bikes':
          bucketName = _bikesBucket;
          break;
        default:
          return false;
      }
      
      await Supabase.instance.client.storage
          .from(bucketName)
          .remove([fileName]);
      return true;
    } catch (e) {
      return false;
    }
  }

  /// Public URL alma
  String getPublicUrl(String fileName, String bucketType) {
    String bucketName;
    switch (bucketType) {
      case 'profile':
        bucketName = _profileBucket;
        break;
      case 'cover':
        bucketName = _coverBucket;
        break;
      case 'events':
        bucketName = _eventsBucket;
        break;
      case 'groups':
        bucketName = _groupsBucket;
        break;
      case 'posts':
        bucketName = _postsBucket;
        break;
      case 'bikes':
        bucketName = _bikesBucket;
        break;
      default:
        bucketName = _profileBucket;
    }
    
    return Supabase.instance.client.storage
        .from(bucketName)
        .getPublicUrl(fileName);
  }

  /// Mevcut kullanıcı ID'sini al
  Future<String?> _getCurrentUserId() async {
    try {
      // Önce Supabase auth'dan al
      final supabaseUser = Supabase.instance.client.auth.currentUser;
      if (supabaseUser?.id != null) {
        return supabaseUser!.id;
      }
      
      // Supabase'den alamazsak backend'den al
      final username = await ServiceLocator.user.getCurrentUsername();
      if (username != null) {
        // Backend'den kullanıcı ID'sini al
        final response = await ServiceLocator.api.get('users/$username/profile/');
        if (response.statusCode == 200 && response.data != null) {
          final userData = response.data;
          return userData['id']?.toString();
        }
      }
      
      return null;
    } catch (e) {
      print('🔥 User ID alma hatası: $e');
      return null;
    }
  }

  /// Eski profil fotoğrafını sil
  Future<void> _deleteOldProfilePicture(String userId, String bucket) async {
    try {
      // Kullanıcının profil fotoğrafı klasöründeki tüm dosyaları listele
      final files = await Supabase.instance.client.storage
          .from(bucket)
          .list(path: 'users/$userId/');
      
      // Profil fotoğraflarını bul ve sil
      for (final file in files) {
        if (file.name.startsWith('profile_')) {
          try {
            await Supabase.instance.client.storage
                .from(bucket)
                .remove(['users/$userId/${file.name}']);
            print('🔥 Eski profil fotoğrafı silindi: ${file.name}');
          } catch (e) {
            print('🔥 Profil fotoğrafı silme hatası: $e');
          }
        }
      }
    } catch (e) {
      print('🔥 Eski profil fotoğrafı listeleme hatası: $e');
    }
  }

  /// Eski kapak fotoğrafını sil
  Future<void> _deleteOldCoverPicture(String userId, String bucket) async {
    try {
      // Kullanıcının kapak fotoğrafı klasöründeki tüm dosyaları listele
      final files = await Supabase.instance.client.storage
          .from(bucket)
          .list(path: 'users/$userId/');
      
      // Kapak fotoğraflarını bul ve sil
      for (final file in files) {
        if (file.name.startsWith('cover_')) {
          try {
            await Supabase.instance.client.storage
                .from(bucket)
                .remove(['users/$userId/${file.name}']);
            print('🔥 Eski kapak fotoğrafı silindi: ${file.name}');
          } catch (e) {
            print('🔥 Kapak fotoğrafı silme hatası: $e');
          }
        }
      }
    } catch (e) {
      print('🔥 Eski kapak fotoğrafı listeleme hatası: $e');
    }
  }

  /// Bir sonraki event klasör numarasını al
  Future<int> _getNextEventFolderNumber() async {
    try {
      // events klasöründeki tüm alt klasörleri listele
      final files = await Supabase.instance.client.storage
          .from(_eventsBucket)
          .list(path: 'events/');
      
      print('🔥 Events klasöründeki dosyalar: ${files.map((f) => f.name).toList()}');
      
      // Sadece klasörleri filtrele (numaralı olanlar)
      final folders = files.where((file) => 
        file.name != null && 
        RegExp(r'^\d+$').hasMatch(file.name)
      ).toList();
      
      print('🔥 Numaralı klasörler: ${folders.map((f) => f.name).toList()}');
      
      if (folders.isEmpty) {
        // İlk klasör: 5 (3 ve 4 zaten mevcut olduğu için)
        return 5;
      }
      
      // Mevcut klasör numaralarını al ve en büyüğünü bul
      final numbers = folders.map((folder) => int.tryParse(folder.name ?? '0') ?? 0).toList();
      final maxNumber = numbers.isEmpty ? 0 : numbers.reduce((a, b) => a > b ? a : b);
      
      // Bir sonraki numara
      return maxNumber + 1;
    } catch (e) {
      print('🔥 Event klasör numarası alma hatası: $e');
      // Hata durumunda 5'ten başla
      return 5;
    }
  }

  /// Bir sonraki grup klasör numarasını al
  Future<int> _getNextGroupFolderNumber() async {
    try {
      // groups klasöründeki tüm alt klasörleri listele
      final files = await Supabase.instance.client.storage
          .from(_groupsBucket)
          .list(path: 'groups/');
      
      print('🔥 Groups klasöründeki dosyalar: ${files.map((f) => f.name).toList()}');
      
      // Sadece klasörleri filtrele (numaralı olanlar)
      final folders = files.where((file) => 
        file.name != null && 
        RegExp(r'^\d+$').hasMatch(file.name)
      ).toList();
      
      print('🔥 Numaralı grup klasörleri: ${folders.map((f) => f.name).toList()}');
      
      if (folders.isEmpty) {
        // İlk klasör: 1
        return 1;
      }
      
      // Mevcut klasör numaralarını al ve en büyüğünü bul
      final numbers = folders.map((folder) => int.tryParse(folder.name ?? '0') ?? 0).toList();
      final maxNumber = numbers.isEmpty ? 0 : numbers.reduce((a, b) => a > b ? a : b);
      
      // Bir sonraki numara
      return maxNumber + 1;
    } catch (e) {
      print('🔥 Group klasör numarası alma hatası: $e');
      // Hata durumunda 1'den başla
      return 1;
    }
  }
}

File: backend/users/services/supabase_storage_service.py
"""
Supabase Storage Service
Profil fotoğrafı ve kapak fotoğrafı yükleme için Supabase Storage kullanımı
"""
import os
import logging
from django.conf import settings
from supabase import create_client, Client
from typing import Optional, Dict, Any

logger = logging.getLogger(__name__)

class SupabaseStorageService:
    def __init__(self):
        self.client: Optional[Client] = None
        self.is_available = False
        
        try:
            self.supabase_url = getattr(settings, 'SUPABASE_URL', None)
            self.supabase_service_key = getattr(settings, 'SUPABASE_SERVICE_KEY', None)
            self.profile_bucket = getattr(settings, 'SUPABASE_BUCKET', 'profile_pictures')
            self.cover_bucket = getattr(settings, 'SUPABASE_COVER_BUCKET', 'cover_pictures')
            self.events_bucket = getattr(settings, 'SUPABASE_EVENTS_BUCKET', 'events_pictures')
            
            print(f"=== SUPABASE STORAGE SERVICE INIT ===")
            print(f"SUPABASE_URL: {self.supabase_url}")
            print(f"SUPABASE_SERVICE_KEY: {'VAR' if self.supabase_service_key else 'YOK'}")
            print(f"events_bucket: {self.events_bucket}")
            
            if self.supabase_url and self.supabase_service_key:
                self.client = create_client(self.supabase_url, self.supabase_service_key)
                self.is_available = True
                print("✅ Supabase Storage servisi başarıyla başlatıldı")
                
                # Bucket'ları kontrol et ve oluştur
                self._ensure_buckets_exist()
                
                logger.info("Supabase Storage servisi başarıyla başlatıldı")
            else:
                print("❌ Supabase Storage credentials eksik")
                print(f"URL var mı: {bool(self.supabase_url)}")
                print(f"SERVICE_KEY var mı: {bool(self.supabase_service_key)}")
                logger.warning("Supabase Storage credentials eksik")
                
        except Exception as e:
            print(f"❌ Supabase Storage servisi başlatılamadı: {e}")
            logger.error(f"Supabase Storage servisi başlatılamadı: {e}")
            self.is_available = False

    def _ensure_buckets_exist(self):
        """Gerekli bucket'ların var olduğundan emin ol"""
        try:
            print("=== BUCKET KONTROL VE OLUŞTURMA ===")
            
            # Mevcut bucket'ları listele
            existing_buckets = self.client.storage.list_buckets()
            existing_bucket_names = [bucket.name for bucket in existing_buckets]
            print(f"Mevcut bucket'lar: {existing_bucket_names}")
            
            # Gerekli bucket'lar
            required_buckets = [
                self.profile_bucket,
                self.cover_bucket, 
                self.events_bucket
            ]
            
            for bucket_name in required_buckets:
                if bucket_name not in existing_bucket_names:
                    print(f"Bucket oluşturuluyor: {bucket_name}")
                    try:
                        self.client.storage.create_bucket(bucket_name, public=True)
                        print(f"✅ Bucket oluşturuldu: {bucket_name}")
                    except Exception as e:
                        print(f"❌ Bucket oluşturulamadı {bucket_name}: {e}")
                else:
                    print(f"✅ Bucket mevcut: {bucket_name}")
                    
        except Exception as e:
            print(f"❌ Bucket kontrol hatası: {e}")

    def upload_profile_picture(self, file, username: str) -> Dict[str, Any]:
        """
        Profil fotoğrafını Supabase Storage'a yükler
        """
        if not self.is_available:
            return {
                'success': False,
                'error': 'Supabase Storage servisi kullanılamıyor'
            }
        
        try:
            # Dosya adını oluştur
            file_extension = file.name.split('.')[-1] if '.' in file.name else 'jpg'
            file_name = f"{username}/profile_{username}_{os.urandom(4).hex()}.{file_extension}"
            
            # Dosyayı yükle
            result = self.client.storage.from_(self.profile_bucket).upload(
                file_name,
                file.read(),
                file_options={
                    "content-type": file.content_type,
                    "upsert": True  # Aynı isimde dosya varsa üzerine yaz
                }
            )
            
            if result:
                # Public URL'i al
                public_url = self.client.storage.from_(self.profile_bucket).get_public_url(file_name)
                
                logger.info(f"Profil fotoğrafı başarıyla yüklendi: {file_name}")
                return {
                    'success': True,
                    'url': public_url,
                    'file_name': file_name
                }
            else:
                return {
                    'success': False,
                    'error': 'Dosya yükleme başarısız'
                }
                
        except Exception as e:
            logger.error(f"Profil fotoğrafı yükleme hatası: {e}")
            return {
                'success': False,
                'error': f'Dosya yükleme hatası: {str(e)}'
            }

    def upload_cover_picture(self, file, username: str) -> Dict[str, Any]:
        """
        Kapak fotoğrafını Supabase Storage'a yükler
        """
        if not self.is_available:
            return {
                'success': False,
                'error': 'Supabase Storage servisi kullanılamıyor'
            }
        
        try:
            # Dosya adını oluştur
            file_extension = file.name.split('.')[-1] if '.' in file.name else 'jpg'
            file_name = f"{username}/cover_{username}_{os.urandom(4).hex()}.{file_extension}"
            
            # Dosyayı yükle
            result = self.client.storage.from_(self.cover_bucket).upload(
                file_name,
                file.read(),
                file_options={
                    "content-type": file.content_type,
                    "upsert": True  # Aynı isimde dosya varsa üzerine yaz
                }
            )
            
            if result:
                # Public URL'i al
                public_url = self.client.storage.from_(self.cover_bucket).get_public_url(file_name)
                
                logger.info(f"Kapak fotoğrafı başarıyla yüklendi: {file_name}")
                return {
                    'success': True,
                    'url': public_url,
                    'file_name': file_name
                }
            else:
                return {
                    'success': False,
                    'error': 'Dosya yükleme başarısız'
                }
                
        except Exception as e:
            logger.error(f"Kapak fotoğrafı yükleme hatası: {e}")
            return {
                'success': False,
                'error': f'Dosya yükleme hatası: {str(e)}'
            }

    def upload_event_picture(self, file, event_id: str) -> Dict[str, Any]:
        """
        Event kapak fotoğrafını Supabase Storage'a yükler
        """
        print(f"=== SUPABASE STORAGE UPLOAD BAŞLADI ===")
        print(f"is_available: {self.is_available}")
        print(f"events_bucket: {self.events_bucket}")
        print(f"event_id: {event_id}")
        
        if not self.is_available:
            print("❌ Supabase Storage servisi kullanılamıyor")
            return {
                'success': False,
                'error': 'Supabase Storage servisi kullanılamıyor'
            }
        
        try:
            # Dosya adını oluştur
            file_extension = file.name.split('.')[-1] if '.' in file.name else 'jpg'
            file_name = f"events/{event_id}/cover_{event_id}_{os.urandom(4).hex()}.{file_extension}"
            print(f"Dosya adı oluşturuldu: {file_name}")
            print(f"Dosya boyutu: {file.size} bytes")
            print(f"Content-Type: {file.content_type}")
            
            # Dosyayı yükle - events_bucket kullan
            print(f"Supabase'e yükleme başlıyor...")
            result = self.client.storage.from_(self.events_bucket).upload(
                file_name,
                file.read(),
                file_options={
                    "content-type": file.content_type,
                    "upsert": True  # Aynı isimde dosya varsa üzerine yaz
                }
            )
            print(f"Upload result: {result}")
            
            if result:
                # Public URL'i al - events_bucket kullan
                public_url = self.client.storage.from_(self.events_bucket).get_public_url(file_name)
                print(f"Public URL oluşturuldu: {public_url}")
                
                logger.info(f"Event kapak fotoğrafı başarıyla yüklendi: {file_name}")
                return {
                    'success': True,
                    'url': public_url,
                    'file_name': file_name
                }
            else:
                print("❌ Upload result False döndü")
                return {
                    'success': False,
                    'error': 'Dosya yükleme başarısız'
                }
                
        except Exception as e:
            print(f"❌ Exception oluştu: {str(e)}")
            logger.error(f"Event kapak fotoğrafı yükleme hatası: {e}")
            return {
                'success': False,
                'error': f'Dosya yükleme hatası: {str(e)}'
            }

    def delete_file(self, bucket: str, file_name: str) -> bool:
        """
        Supabase Storage'dan dosya siler
        """
        if not self.is_available:
            return False
        
        try:
            result = self.client.storage.from_(bucket).remove([file_name])
            logger.info(f"Dosya silindi: {bucket}/{file_name}")
            return True
        except Exception as e:
            logger.error(f"Dosya silme hatası: {e}")
            return False
